File: car/app/app-samples/showcase/common/src/main/java/androidx/car/app/sample/showcase/common/screens/templatelayouts/gridtemplates/GridTemplateDemoScreen.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.car.app.sample.showcase.common.screens.templatelayouts.gridtemplates;

import static androidx.car.app.CarToast.LENGTH_SHORT;
import static androidx.car.app.model.Action.BACK;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.os.Looper;

import androidx.car.app.CarContext;
import androidx.car.app.CarToast;
import androidx.car.app.Screen;
import androidx.car.app.constraints.ConstraintManager;
import androidx.car.app.model.Action;
import androidx.car.app.model.CarIcon;
import androidx.car.app.model.GridItem;
import androidx.car.app.model.GridTemplate;
import androidx.car.app.model.Header;
import androidx.car.app.model.ItemList;
import androidx.car.app.model.OnClickListener;
import androidx.car.app.model.Template;
import androidx.car.app.sample.showcase.common.R;
import androidx.car.app.versioning.CarAppApiLevels;
import androidx.core.graphics.drawable.IconCompat;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;

import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

/** Creates a screen that demonstrates usage of the full screen {@link GridTemplate}. */
public final class GridTemplateDemoScreen extends Screen implements DefaultLifecycleObserver {
    private static final int MAX_GRID_ITEMS = 100;
    private static final int LOADING_TIME_MILLIS = 2000;
    private final Handler mHandler = new Handler(Looper.getMainLooper());

    private @Nullable IconCompat mImage;
    private @Nullable IconCompat mIcon;
    private boolean mIsFourthItemLoading;
    private boolean mThirdItemToggleState;
    private boolean mFourthItemToggleState;
    private boolean mFifthItemToggleState;

    public GridTemplateDemoScreen(@NonNull CarContext carContext) {
        super(carContext);
        getLifecycle().addObserver(this);
        mIsFourthItemLoading = false;
        mThirdItemToggleState = false;
        mFourthItemToggleState = true;
        mFifthItemToggleState = false;
    }

    @Override
    public void onCreate(@NonNull LifecycleOwner owner) {
        Resources resources = getCarContext().getResources();
        Bitmap bitmap = BitmapFactory.decodeResource(resources, R.drawable.test_image_square);
        mImage = IconCompat.createWithBitmap(bitmap);
        mIcon = IconCompat.createWithResource(getCarContext(), R.drawable.ic_fastfood_white_48dp);
    }

    @Override
    @SuppressWarnings({"FutureReturnValueIgnored"})
    public void onStart(@NonNull LifecycleOwner owner) {
        mIsFourthItemLoading = false;

        // Post a message that starts loading the fourth item for some time.
        triggerFourthItemLoading();
    }

    private GridItem createGridItem(int index) {
        switch (index) {
            case 0:
                // Grid item with an icon and a title.
                return createGridItem(createCarIconFromImage(mIcon), GridItem.IMAGE_TYPE_ICON,
                        getTextStringFromId(R.string.non_actionable));
            case 1:
                // Grid item with an icon, a title, onClickListener and no text.
                return createGridItem(createCarIconFromImage(mIcon), GridItem.IMAGE_TYPE_ICON,
                        getTextStringFromId(R.string.second_item), createOnClickListener(
                                (String) getTextStringFromId(R.string.second_item_toast_msg),
                                LENGTH_SHORT));

            case 2:
                // Grid item with an icon marked as icon, a title, a text and a toggle in
                // unchecked state.
                return createGridItem(createCarIconFromImage(mIcon),
                        GridItem.IMAGE_TYPE_LARGE, getTextStringFromId(R.string.third_item),
                        mThirdItemToggleState
                                ? getTextStringFromId(R.string.checked_action_title) :
                                getTextStringFromId(R.string.unchecked_action_title),
                        createOnClickListenerForThirdItem());

            case 3:
                // Grid item with an image, a title, a long text and a toggle that takes some
                // time to
                // update.
                if (mIsFourthItemLoading) {
                    return createGridItem(getTextStringFromId(R.string.fourth_item),
                            mFourthItemToggleState
                                    ? getTextStringFromId(R.string.on_action_title) :
                                    getTextStringFromId(R.string.off_action_title),
                            true);
                } else {
                    return createGridItem(createCarIconFromImage(mImage),
                            getTextStringFromId(R.string.fourth_item),
                            mFourthItemToggleState
                                    ? getTextStringFromId(R.string.on_action_title) :
                                    getTextStringFromId(R.string.off_action_title),
                            this::triggerFourthItemLoading);
                }
            case 4:
                // Grid item with a large image, a long title, no text and a toggle in unchecked
                // state.
                return createGridItem(createCarIconFromImage(mIcon),
                        GridItem.IMAGE_TYPE_LARGE, getTextStringFromId(R.string.fifth_item),
                        createOnClickListenerForFifthItem());
            case 5:
                // Grid item with an image marked as an icon, a long title, a long text and
                // onClickListener.
                return createGridItem(createCarIconFromImage(mIcon),
                        GridItem.IMAGE_TYPE_ICON, getTextStringFromId(R.string.sixth_item),
                        getTextStringFromId(R.string.sixth_item),
                        createOnClickListener(
                                (String) getTextStringFromId(R.string.sixth_item_toast_msg),
                                LENGTH_SHORT));

            default:
                String titleText = (index + 1) + "th item";
                String toastText = "Clicked " + (index + 1) + "th item";

                return createGridItem(createCarIconFromImage(mIcon),
                        GridItem.IMAGE_TYPE_ICON, titleText, createOnClickListener(toastText,
                                LENGTH_SHORT));
        }
    }


    @Override
    public @NonNull Template onGetTemplate() {
        int itemLimit = 6;
        // Adjust the item limit according to the car constrains.
        if (getCarContext().getCarAppApiLevel() > CarAppApiLevels.LEVEL_1) {
            itemLimit =
                    Math.min(MAX_GRID_ITEMS,
                            getCarContext().getCarService(ConstraintManager.class).getContentLimit(
                                    ConstraintManager.CONTENT_LIMIT_TYPE_GRID));
        }

        ItemList.Builder gridItemListBuilder = new ItemList.Builder();
        for (int i = 0; i <= itemLimit; i++) {
            gridItemListBuilder.addItem(createGridItem(i));
        }

        Action settings = new Action.Builder()
                .setTitle(getCarContext().getString(
                        R.string.settings_action_title))
                .setOnClickListener(
                        () -> CarToast.makeText(
                                        getCarContext(),
                                        getCarContext().getString(R.string.settings_toast_msg),
                                        LENGTH_SHORT)
                                .show())
                .build();
        return new GridTemplate.Builder()
                .setHeader(new Header.Builder()
                        .setStartHeaderAction(BACK)
                        .setTitle(getCarContext().getString(R.string.grid_template_demo_title))
                        .addEndHeaderAction(settings)
                        .build())
                .setSingleList(gridItemListBuilder.build())
                .build();
    }

    /**
     * Changes the fourth item to a loading state for some time and changes it back to the loaded
     * state.
     */
    private void triggerFourthItemLoading() {
        mHandler.post(
                () -> {
                    mIsFourthItemLoading = true;
                    invalidate();

                    mHandler.postDelayed(
                            () -> {
                                mIsFourthItemLoading = false;
                                mFourthItemToggleState = !mFourthItemToggleState;
                                invalidate();
                            },
                            LOADING_TIME_MILLIS);
                });
    }

    /**
     * Custom listener for third grid item
     */
    private OnClickListener createOnClickListenerForThirdItem() {
        return () -> {
            mThirdItemToggleState = !mThirdItemToggleState;
            CarToast.makeText(
                            getCarContext(),
                            getCarContext().getString(
                                    R.string.third_item_checked_toast_msg)
                                    + ": " + mThirdItemToggleState,
                            LENGTH_SHORT)
                    .show();
            invalidate();
        };
    }

    /**
     * Custom listener for fifth grid item
     */
    private OnClickListener createOnClickListenerForFifthItem() {
        return () -> {
            mFifthItemToggleState = !mFifthItemToggleState;
            CarToast.makeText(
                            getCarContext(),
                            getCarContext().getString(
                                    R.string.fifth_item_checked_toast_msg)
                                    + ": "
                                    + mFifthItemToggleState,
                            LENGTH_SHORT)
                    .show();
            invalidate();
        };
    }

    /**
     * Returns CharSequence text for given id
     */
    private CharSequence getTextStringFromId(int id) {
        return getCarContext().getString(id);
    }

    /**
     * Create an onClickListener using a Car Toast
     */
    private OnClickListener createOnClickListener(String toastText, int toastLength) {
        return () ->
                CarToast.makeText(
                                getCarContext(),
                                toastText,
                                toastLength)
                        .show();
    }

    /**
     * Creates CarIcon from given IconCompat image
     */
    private CarIcon createCarIconFromImage(IconCompat image) {
        return new CarIcon.Builder(image).build();
    }

    /**
     * A number of overloaded constructors with different parameters to build grid items
     */
    private GridItem createGridItem(CharSequence title, CharSequence text,
            boolean isLoading) {
        return new GridItem.Builder()
                .setTitle(title)
                .setText(text)
                .setLoading(isLoading)
                .build();
    }

    private GridItem createGridItem(CarIcon carIcon,
            int imageType, CharSequence title, CharSequence text,
            OnClickListener onClickListener) {
        return new GridItem.Builder()
                .setImage(carIcon, imageType)
                .setTitle(title)
                .setText(text)
                .setOnClickListener(onClickListener)
                .build();
    }

    private GridItem createGridItem(CarIcon carIcon,
            CharSequence title, CharSequence text,
            OnClickListener onClickListener) {
        return new GridItem.Builder()
                .setImage(carIcon)
                .setTitle(title)
                .setText(text)
                .setOnClickListener(onClickListener)
                .build();
    }

    private GridItem createGridItem(CarIcon carIcon,
            int imageType, CharSequence title,
            OnClickListener onClickListener) {
        return new GridItem.Builder()
                .setImage(carIcon, imageType)
                .setTitle(title)
                .setOnClickListener(onClickListener)
                .build();
    }

    private GridItem createGridItem(CarIcon carIcon,
            int imageType, CharSequence title) {
        return new GridItem.Builder()
                .setImage(carIcon, imageType)
                .setTitle(title)
                .build();
    }
}

File: car/app/app-samples/showcase/common/src/main/java/androidx/car/app/sample/showcase/common/screens/templatelayouts/gridtemplates/GridTemplateDemoScreen.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.car.app.sample.showcase.common.screens.templatelayouts.gridtemplates

import android.os.Handler
import android.os.Looper
import androidx.annotation.OptIn
import androidx.car.app.CarContext
import androidx.car.app.Screen
import androidx.car.app.annotations.ExperimentalCarApi
import androidx.car.app.constraints.ConstraintManager.CONTENT_LIMIT_TYPE_GRID
import androidx.car.app.model.Action
import androidx.car.app.model.Action.BACK
import androidx.car.app.model.GridItem
import androidx.car.app.model.GridTemplate
import androidx.car.app.model.Header
import androidx.car.app.model.ItemList
import androidx.car.app.model.Template
import androidx.car.app.navigation.model.MapWithContentTemplate
import androidx.car.app.sample.showcase.common.R
import androidx.car.app.sample.showcase.common.utils.CarContextAware
import androidx.car.app.sample.showcase.common.utils.createGridItem
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner

/** Creates a screen that demonstrates usage of the full screen [GridTemplate]. */
@OptIn(ExperimentalCarApi::class)
class GridTemplateDemoScreen(carContext: CarContext) :
    Screen(carContext), CarContextAware, DefaultLifecycleObserver {

    private val handler = Handler(Looper.getMainLooper())
    private val itemLimit = determineListLimit(MAX_GRID_ITEMS, CONTENT_LIMIT_TYPE_GRID)
    private val pageImage = getCarIcon(R.drawable.test_image_square)
    private val pageIcon = getCarIcon(R.drawable.ic_fastfood_white_48dp)
    private val mapXIcon = getCarIcon(R.drawable.ic_emoji_food_beverage_white_48dp)
    private val pageTitle = getString(R.string.grid_template_demo_title)
    private val firstItemTitle = getString(R.string.non_actionable)
    private val secondItemTitle = getString(R.string.second_item)
    private val secondItemToastMessage = getString(R.string.second_item_toast_msg)
    private val thirdItemTitle = getString(R.string.third_item)
    private val thirdItemToastMessage = getString(R.string.third_item_checked_toast_msg)
    private val fourthItemTitle = getString(R.string.fourth_item)
    private val fifthItemTitle = getString(R.string.fifth_item)
    private val fifthItemToastMessage = getString(R.string.fifth_item_checked_toast_msg)
    private val sixthItemTitle = getString(R.string.sixth_item)
    private val sixthItemToastMessage = getString(R.string.sixth_item_toast_msg)
    private val checkedText = getString(R.string.checked_action_title)
    private val uncheckedText = getString(R.string.unchecked_action_title)
    private val stateTextOn = getString(R.string.on_action_title)
    private val stateTextOff = getString(R.string.off_action_title)
    private var isFourthItemLoading = false
    private var thirdItemChecked = false
    private var fourthItemChecked = true
    private var fifthItemChecked = false

    init {
        lifecycle.addObserver(this)
    }

    companion object {
        private const val MAX_GRID_ITEMS = 100
        private const val LOADING_TIME_MILLIS = 2000
    }

    override fun onStart(owner: LifecycleOwner) {
        isFourthItemLoading = false

        // Post a message that starts loading the fourth item for some time.
        triggerFourthItemLoading()
    }

    private fun makeGridItem(index: Int): GridItem =
        when (index) {
            0 -> buildFirstGridItem()
            1 -> buildSecondGridItem()
            2 -> buildThirdGridItem()
            3 -> buildFourthGridItem()
            4 -> buildFifthGridItem()
            5 -> buildSixthGridItem()
            else -> buildDefaultGridItem(index)
        }

    /** Grid item 0: Non-actionable item with icon and title. */
    private fun buildFirstGridItem(): GridItem =
        createGridItem(title = firstItemTitle, image = pageIcon)

    /** Grid item 1: Actionable item with icon, title, and toast click listener. */
    private fun buildSecondGridItem(): GridItem =
        createGridItem(
            title = secondItemTitle,
            image = pageIcon,
            clickListener = { makeToast(secondItemToastMessage).show() },
        )

    /** Grid item 2: Interactive item with toggle state text. */
    private fun buildThirdGridItem(): GridItem =
        createGridItem(
            title = thirdItemTitle,
            image = pageIcon,
            text = if (thirdItemChecked) checkedText else uncheckedText,
            clickListener = {
                thirdItemChecked = !thirdItemChecked
                makeToast("$thirdItemToastMessage: $thirdItemChecked").show()
                invalidate()
            },
        )

    /** Grid item 3: Item with simulated 2-second async loading state. */
    private fun buildFourthGridItem(): GridItem {
        val stateText = if (fourthItemChecked) stateTextOn else stateTextOff
        return if (isFourthItemLoading) {
            createGridItem(title = fourthItemTitle, text = stateText, isLoading = true)
        } else {
            createGridItem(
                title = fourthItemTitle,
                text = stateText,
                image = pageImage,
                clickListener = ::triggerFourthItemLoading,
            )
        }
    }

    /**
     * Changes the fourth item to a loading state for some time and changes it back to the loaded
     * state.
     */
    private fun triggerFourthItemLoading() {
        handler.post {
            isFourthItemLoading = true
            invalidate()

            handler.postDelayed(
                {
                    isFourthItemLoading = false
                    fourthItemChecked = !fourthItemChecked
                    invalidate()
                },
                LOADING_TIME_MILLIS.toLong(),
            )
        }
    }

    /** Grid item 4: Interactive item with checked/unchecked toast state. */
    private fun buildFifthGridItem(): GridItem =
        createGridItem(
            title = fifthItemTitle,
            image = pageIcon,
            clickListener = {
                fifthItemChecked = !fifthItemChecked
                makeToast("$fifthItemToastMessage: $fifthItemChecked").show()
                invalidate()
            },
        )

    /** Grid item 5: Item with icon, long title, long text, and toast click listener. */
    private fun buildSixthGridItem(): GridItem =
        createGridItem(
            title = sixthItemTitle,
            text = sixthItemTitle,
            image = pageIcon,
            clickListener = { makeToast(sixthItemToastMessage).show() },
        )

    /** Default grid item fallback for indices >= 6. */
    private fun buildDefaultGridItem(index: Int): GridItem =
        createGridItem(
            title = "${index + 1}th item",
            image = pageIcon,
            clickListener = { makeToast("Clicked ${index + 1}th item").show() },
        )

    override fun onGetTemplate(): Template {
        return buildGridTemplate()
    }

    /** Helper method to build the GridTemplate. */
    private fun buildGridTemplate(): GridTemplate {
        val gridItemList =
            ItemList.Builder()
                .apply { repeat(itemLimit) { index -> addItem(makeGridItem(index)) } }
                .build()

        val mapXAction =
            Action.Builder()
                .setTitle("Map+X this!")
                .setIcon(mapXIcon)
                .setOnClickListener { screenManager.push(MapGridDemoScreen(carContext)) }
                .build()

        val pageHeader =
            Header.Builder()
                .setStartHeaderAction(BACK)
                .setTitle(pageTitle)
                .addEndHeaderAction(mapXAction)
                .build()

        return GridTemplate.Builder().setHeader(pageHeader).setSingleList(gridItemList).build()
    }

    /**
     * A new screen that displays the MapWithContentTemplate containing the exact same GridTemplate.
     */
    private inner class MapGridDemoScreen(carContext: CarContext) : Screen(carContext) {

        override fun onGetTemplate(): Template {
            val innerTemplate = this@GridTemplateDemoScreen.buildGridTemplate()
            return MapWithContentTemplate.Builder().setContentTemplate(innerTemplate).build()
        }
    }
}
